src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c

```diff
@@ -5339,41 +5367,92 @@ void wm_vuldet_feed_update_by_interval(vu_task_ctx_t* current_task, wm_vuldet_t*
+                f_agent = agent_aux;
+            }
+            agents = agent_aux;
+            agents->next = NULL;
         }
-    }
+        i++;
+    } while(agents_array[i] != OS_INVALID);
+    os_free(agents_array);
 
     // Scanning agents
-    agents_array = wdb_get_agents_by_connection_status(AGENT_CS_ACTIVE, &sock);
-    while(agents_array && agents_array[i] != OS_INVALID && current_task && WM_TASK_CANCELLED != current_task->task_status) {
-        if (wm_vuldet_collect_agent_info(agents_array[i], vuldet)) {
-            mterror(WM_VULNDETECTOR_LOGTAG, VU_NO_AGENT_ERROR);
-        } else {
-            snprintf(status_message, OS_SIZE_64, " Agent: %.3d.", agents_array[i]);
+    bool retry_agents = false;
+    bool abort_scan = false;
+    char status_message[OS_SIZE_64] = {0};
+
+    mtinfo(WM_VULNDETECTOR_LOGTAG, VU_START_SCAN);
+    do {
+        retry_agents = false;
+        time_t first_fail_scan = 0;
+        agents = f_agent;
+        while (agents && current_task && WM_TASK_CANCELLED != current_task->task_status) {
+            vuldet->agent_to_scan = agents;
+            snprintf(status_message, OS_SIZE_64, " Agent: %.3d.", atoi(agents->agent_id));
             wm_vuldet_create_and_send_message(current_task, WM_TASK_IN_PROGRESS, status_message);
-            if (wm_vuldet_run_scan(vuldet)) {
+
+            if (wm_vuldet_check_agent_vulnerabilities(vuldet)) {
                 mterror(WM_VULNDETECTOR_LOGTAG, VU_AG_CHECK_ERR);
+                abort_scan = true;
+                break;
+            } else {
+                if (agents->pending_attempts > 0) {
+                    retry_agents = true;
+                    if (!first_fail_scan) {
+                        first_fail_scan = time(NULL);
+                    }
+                    agents = agents->next;
+                } else {
+                    agent_aux = agents->next;
+                    if (f_agent->agent_id == agents->agent_id) {
+                        f_agent = agent_aux;
+                    }
+                    wm_vuldet_free_scan_agent(agents);
+                    agents = agent_aux;
+                }
             }
+            current_task = (vu_task_ctx_t*) linked_queue_read_ex(vu_tasks_queue);
         }
-        i++;
-        current_task = (vu_task_ctx_t*) linked_queue_read_ex(vu_tasks_queue);
-    }
+
+        if (retry_agents && !abort_scan && WM_TASK_CANCELLED != current_task->task_status) {
+            time_t sleep_time = first_fail_scan + vuldet->retry_interval - time(NULL);
+            if (sleep_time <= 0) {
+                sleep_time = 1;
+            }
+            mtdebug1(WM_VULNDETECTOR_LOGTAG, "Going to sleep %ld seconds before retrying pending agents", sleep_time);
+            sleep (sleep_time);
+        }
+    } while (retry_agents && !abort_scan && WM_TASK_CANCELLED != current_task->task_status);
+    wm_vuldet_clean_all_scan_agent_structures(f_agent);
+
     mtinfo(WM_VULNDETECTOR_LOGTAG, VU_END_SCAN);
```

Comment:
You are right.
It is made this way because in the case of a scan by interval is really clear when we start with all the agents and when we end with them.

But in case of a scan on demand, I thought that 
- It wouldn't be correct to print the scan started/finished log for every agent
- I don't know how many agents we have or the order I'll receive them
- We already have the log **VU_START_AG_AN**